### DISCUSSION TITLE
Operators revisited

### DISCUSSION THREAD

**Discussion by 2colours** [perl6/doc#4337] [Ideas]
Hello,

since operators are one of my favorite topics (and also a topic that involves many misunderstandings and documentation issues), I decided to open this brainstorming session about possible directions to rework documentation involving operators.

Quick disclaimer: when I use the word "productive", I usually mean the linguistic sense: the ability to create previously undefined concepts (here, apparent operators) via regular deduction.

## Issues and challenges so far

To get the hang of it, I went through all open issues in the repo, to collect some thoughts. I'm adding a little "tldr" to all of them.

https://github.com/Raku/doc/issues/4160 -> `how do we relate to the "decont operator" that isn't even standalone syntax, just a mental framework?`
https://github.com/Raku/doc/issues/4058 -> `what can/should we say about _intermediary_ computation in an expression? how to glue it to the evaluation model?`
https://github.com/Raku/doc/issues/4073 -> `how to better distinguish the two identically-looking assignments that otherwise can have different properties?`
https://github.com/Raku/doc/issues/2861 -> `how to improve the op= situation (and also the meta-operator situation) so that these synthetic operators will be easier to look up and understand?`
https://github.com/Raku/doc/issues/2337 -> `how to define and categorize operators and related content; productivity and custom definition concerns`
https://github.com/Raku/doc/issues/2129 -> `bring in identity value as a systematically described aspect of all operators`
https://github.com/Raku/doc/issues/1730 -> `customizability of user-defined operators`
https://github.com/Raku/doc/issues/4033 -> `what can/should we say about the underlying behavior of core operators; how can we establish better communication than "not a real operator"?`

The following issues are also "operator issues" but I think they are merely content issues; nothing systemic about them:

https://github.com/Raku/doc/issues/4326
https://github.com/Raku/doc/issues/4133
https://github.com/Raku/doc/issues/4077 (this is about hyper meta-operators which can still be tough stuff...)
https://github.com/Raku/doc/issues/3839
https://github.com/Raku/doc/issues/3815
https://github.com/Raku/doc/issues/3756 (is this even an operator, but anyway)
https://github.com/Raku/doc/issues/3333 (I think `.=` as an intersection of two special situations will require some attention actually)
https://github.com/Raku/doc/issues/3346
https://github.com/Raku/doc/issues/3189
https://github.com/Raku/doc/issues/3037

## My ideas

Right now, operators mostly reside in a big monolith at https://docs.raku.org/language/operators. I would like to break this page into two: `language/operators` would mostly become a guide to Raku operators, talking about several aspects (associativity, infix/prefix/etc. positions, thunkiness, meta-operators/productivity, underlying behavior, etc.) It would contain rationale for "methodoperators" and "thunkiness" at the very least. To remove the big table and list of operators, we could consider URI redirects (using the #component as well for distinguishing, if that's possible), or we could keep that part of the site, preferably moving it out of the way, at the end of the page.  
The other part would have a better place under Type in my opinion, but if that's unwelcome (since these are not literal "types"), it could still be a separate thing under Language. It would have roughly the same operators that are listed currently but with much more qualitative information for each: again, associativity, thunkiness, composability with meta-operators, underlying behavior, precedence level and whatnot. Of course, then the really good thing would be to optionally also provide qualitative information about operators in a huge table that can be at least ordered - but let's say, that's a big wish for now.

The point is, separate the "guide" purpose and the "reference" purpose and make them grow in their logical ways (the "guide" deserves definitions, examples and rationale about certain concepts; the "reference" needs accurate, "measurable" information easily accessible).

**Comment by doomvox** [comment]
Looking over the language/operators page, it does seem like a bit of an info dump up top, with the useful examples buried down below.  There might be some advantage in trying to list which operators a new user might really want to see documentation for,  (perhaps a synopsis section with some exemplar operators worked through?). 

I note that some things on that page, it wouldn't normally occur to us to think of as operators, e.g. using curly braces to create a Hash or a Block.   That feels more like a language syntax feature than an operator...

**Reply by 2colours** [reply]
Good points!

Personally, an "example-first approach" (if that makes sense?) doesn't really sit with me - but maybe something like snippets for each individual topic ("this is how you can use a meta-operator with a newly defined operator", "this is an example for chain associativity", etc.) can be useful without feeling too much out of context. I think it's not easy to add the right amount of examples and not fall to one of the extremes.

You are absolutely right about the things that appear as operators but you might not notice or expect them to show up, like `[...]` or `{...}`, especially when there are also the "quoting structures" that are apparently not operators, including `<...>`. I have been thinking about this as well and the best idea I have is a sort of cross-referencing; that is, if you look up `List`s, it would be reinforced that they are created using the comma operator, if you look up `Array`s, it would mention that usually they are defined using the `[...]` operator - I think the latter might be "trap-worthy" as well because it's an idempotent operator (didn't look up if it's already there) - and so on. It's realistic and probably not too annoying that with a common type there can be a little description of its usual appearance, "literal" if you will, and that would hopefully increase the discoverability.

**Comment by codesections** [comment]
I really like this idea overall, but I'd like to take it even further.  Right now, [docs.raku.org/reference](https://docs.raku.org/reference) lists around 47 pages and separates them into just two categories: "Fundamental topics" and "General reference".  Looking at "Fundamental topics", there's a huge disparity between just how "fundamental" some of the topics are.  For example [Control flow](https://docs.raku.org/language/control) is genuinely fundamental to reading/writing Raku.  Conversely, the [MOP](https://docs.raku.org/language/mop) or Raku's [newline handling](https://docs.raku.org/language/newline) are … much less fundamental.

What I'd like to see is for the /reference section to be split into several different categories and to have just a *few* pages in a "Fundamental topics" category – those pages that a new Rakoon really ought to read to grasp Raku's fundamentals.

I'm bringing this up because one of the Operators pages you propose sounds like it'd fit in that narrower version of "Fundamental topics" and the other one wouldn't.  If we reorganized along those lines, then I'd really like the idea of having two Operators pages – basically, a "Operators fundamentals" and an "Advanced operators".  But I'm less sold on the idea of having two pages if they're both classified as "Fundamental topics" (not opposed; just less sold – maybe we could pick a name that made the distinction clear).

What do others think about idea of reorganizing /reference into different sub-categories with the goal of getting the list of "fundamental topics" down to a manageable size?

**Reply by codesections** [reply]
(By the way, I'm intentionally avoiding the label "Guides" or "Tutorials" for any of the pages in /reference.  [docs.raku.org/introduction](https://docs.raku.org/introduction) seems like the home for "Guides" or "Tutorials".  But even within /reference, there are some fundamental references, and some references to more advanced topics.

That said, if we *do* reorganize a bit, it might make sense to move some of the pages out of /introduction. some of the "Tutorials" are (appropriately!) fairly advanced; having them live under /introduction feels a bit weird to me.)

**Reply by 2colours** [reply]
I think what you say does make a lot of sense but I feel this will lead us to a dead end again.

The current categorization [is very recent](https://github.com/Raku/doc-website/issues/294). If you read into the comments, I for one already noted that I didn't have the same concept of "reference" in mind. However, previously, it really was just "Language" and an unreasonably promoted "Programs". I think getting categorization right is such a large and broad topic, especially when our hands are tied with the URL's at the very least, that it will lead to endless bikeshedding, no agreement, and based on my mindset, quite possibly people who feel they weren't listened to and that the "whole thing was a waste of time" or "worse than leaving as it was".

Frankly, I believe much more in cross-referencing than some hierarchic structure, especially given the existing conditions. Somebody can find the suitable topic in any way - mostly via search - and then naturally discover anything else that might be related.

All in all, I'd like to be more concerned with the possible content and format of covering the topic of operators here. It's almost a mere technicality that I'd like to split it into two and I'm deliberately using the concepts of "reference" and "guide" here: regardless where we put them, the content should be separated based on this duality. Some people want to learn about operators in general - for them, a big table full of properties is near useless; other people want to check some exact properties of particular operators - for them, a guide is just noise that gets in the way. This is the basic idea - but what should be the content? How to structure it within the two parts?

**Comment by arkiuat** [comment]
Although I like the ideas @codesections presents above, the ones about top-level reorg of the docs as a whole deserve their own separate issue.

At 132K, Language/operators.rakudoc is the longest single document in the tree, and its load time in a browser is annoyingly slower than other pages: annoying because this one is so frequently referred to.  The general discussion of operators at the beginning makes up only about 5% of the whole, so separating it out would make that short section easier to work with, although by itself that wouldn't help with the other 95%.

Really, *any* reasonable scheme of splitting up the remaining 95% into separate sections, each in their own .rakudoc file, would be helpful. All the better if it's a conceptual assist to the process of learning the operators. For instance, I only recently realized (though of course I'd known for a long time, it just hadn't really sunk in) that not all operators are Callables: it might be helpful to separate the operators that are just syntactic sugar for ordinary Callables from the operators that are syntactic sugar, period, with no Callable behind it. I certainly would have found it helpful to have this distinction presented up front, rather than having to gradually figure it out while I was learning the language.

Let me know if you have any ideas how I can help move this issue forward.

**Reply by 2colours** [reply]
Oh yes, the good old "not a real operator", and then the two categories of "synthetic"/"meta" operators (like `//=`) and syntax special-cased operators (most "methodoperators", postcircumfix `< >`).

For the splitting of operators themselves ie. the reference list part, I think there might be a need for a two-layer system, at least I don't know a property that would be a sufficiently good "hash" to really split the operators into portions with managable sizes. In this case, it would make sense for the top layer to be the "position" (how is this called?) of it so prefix, infix, postfix, etc. since that's already an established, very much user-facing categorisation for subroutines. I would probably keep "methodoperators" (in the stricter sense so the ones that indeed have a dot and something on the right handside that isn't quite an expression) in a "miscellaneous" category that they hopefully wouldn't share with too many other things.
Metaoperators could then be approached from two directions: "what-fix" operators they take and "what-fix" the resulting operator is. For example, a reduction meta-operator takes an infix operator (and only that - correct me if I'm wrong) and the result is a prefix operator (if not outright a term, I'm not sure... most of the time, terms and prefix operators are quite similar).
For example, one could write it down at prefix operators that reduce-derived operators are prefix, and describe at infix operators that they can be passed to the reduce meta-operator to obtain a new prefix operator. Or that assignment-derived `foo=` operators can be obtained from infix operators, giving you a new infix operator.

As a general rule of thumb, I don't think it's important to avoid duplication and furthermore, cross-linking is outright a very much welcome thing. If the assignment as a rather complex thing got its own little topic within infix operators, it could (should) mention these synthetic `foo=` assignments but all the same, if meta-operators deserve their own topic, then it could (should) appear there as well.

**Comment by finanalyst** [comment]
There was a discussion in the last Resolver session about re-categorising the existing pages, as mentioned by @codesections . FYI, all sources have a file-level set of meta-data with `:kinds :subkinds :categories`. The different top-level pages, eg. Introduction, Reference, Tutorials, are constructed from the kinds/subkinds/categories meta data. 
Since a new website is gradually being developed, and since we can add any new metadata to the sources, then we can
- have a new set of metadata options, eg, newcomer, tutorial, advanced, or whatever we want. 
- Then a new set of pages can be generated. In fact, we could have multiple overlapping pages.
- The only constraint on the number of top-level pages is how to reference them in the Navigation bar.

This is slightly off-topic here because this thread is about operators in general. However, perhaps we might consider having several operator pages:
- a main one (as is now, which collects all sorts of stuff)
- several sub-pages that reference and explain the different groups of operators
- the sub-pages would not necessarily be 'tutorials' so much as 'explanation / discussion'
